rpc/src/client.rs

```diff
@@ -1,45 +1,150 @@
-pub use self::imp::*;
+//! Clients for the celestia Json-RPC.
+//!
+//! This module aims to provide a convenient way to create a Json-RPC clients. If
+//! you need more configuration options and / or some custom client you can create
+//! one using [`jsonrpsee`] crate directly.
 
 #[cfg(not(target_arch = "wasm32"))]
-mod imp {
-    use crate::Result;
+pub use self::native::Client;
+
+#[cfg(not(target_arch = "wasm32"))]
+mod native {
+    use std::fmt;
+    use std::result::Result as StdResult;
+
+    use crate::{Error, Result};
+    use async_trait::async_trait;
     use http::{header, HeaderValue};
+    use jsonrpsee::core::client::{BatchResponse, ClientT, Subscription, SubscriptionClientT};
+    use jsonrpsee::core::params::BatchRequestBuilder;
+    use jsonrpsee::core::traits::ToRpcParams;
+    use jsonrpsee::core::Error as JrpcError;
     use jsonrpsee::http_client::{HeaderMap, HttpClient, HttpClientBuilder};
     use jsonrpsee::ws_client::{WsClient, WsClientBuilder};
+    use serde::de::DeserializeOwned;
 
-    pub fn new_http(conn_str: &str, auth_token: Option<&str>) -> Result<HttpClient> {
-        let mut headers = HeaderMap::new();
+    /// Json RPC client.
+    pub enum Client {
+        Http(HttpClient),
+        Ws(WsClient),
+    }
 
-        if let Some(token) = auth_token {
-            let val = HeaderValue::from_str(&format!("Bearer {token}"))?;
-            headers.insert(header::AUTHORIZATION, val);
-        }
+    impl Client {
+        /// Create a new Json RPC client.
+        ///
+        /// Only 'http' and 'ws' protocols are supported and they should
+        /// be specified in the provided `conn_str`. For more flexibility
+        /// consider creating the client using [`jsonrpsee`] directly.
+        pub async fn new(conn_str: &str, auth_token: Option<&str>) -> Result<Self> {
+            let mut headers = HeaderMap::new();
+
+            if let Some(token) = auth_token {
+                let val = HeaderValue::from_str(&format!("Bearer {token}"))?;
+                headers.insert(header::AUTHORIZATION, val);
+            }
 
-        let client = HttpClientBuilder::default()
-            .set_headers(headers)
-            .build(conn_str)?;
+            let client = match conn_str.split_once(':') {
+                Some(("http", _)) => Client::Http(
```

Comment:
add `"https"`

Reply:
https and ws are not supported by celestia-node currently

Reply:
Someone can use reverse proxy and have an https in front of it. I think we should support it

Reply:
added